libraries/chain/kv_context_chainbase.cpp

```diff
@@ -0,0 +1,225 @@
+            if (++*it) {
+               have_next = true;
+               break;
+            }
+         }
+         if (!have_next)
+            next_prefix.reset();
+      }
+
+      ~kv_iterator_chainbase() override {}
+
+      bool is_kv_chainbase_context_iterator() const override { return true; }
+
+      template <typename It>
+      kv_it_stat move_to(const It& it) {
+         if (it != idx.end() && it->database_id == database_id && it->contract == contract &&
+             it->kv_key.size() >= prefix.size() && !memcmp(it->kv_key.data(), prefix.data(), prefix.size())) {
+            kv_key.emplace(it->kv_key);
+            return kv_it_stat::iterator_ok;
+         } else {
+            kv_key.reset();
+            return kv_it_stat::iterator_end;
+         }
+      }
+
+      kv_it_stat kv_it_status() override {
+         if (!kv_key)
+            return kv_it_stat::iterator_end;
+         auto* kv = db.find<kv_object, by_kv_key>(boost::make_tuple(database_id, contract, *kv_key));
+         if (kv)
+            return kv_it_stat::iterator_ok;
+         else
+            return kv_it_stat::iterator_erased;
+      }
+
+      int32_t kv_it_compare(const kv_iterator& rhs) override {
+         EOS_ASSERT(rhs.is_kv_chainbase_context_iterator(), kv_bad_iter, "Incompatible key-value iterators");
+         auto& r = static_cast<const kv_iterator_chainbase&>(rhs);
+         EOS_ASSERT(database_id == r.database_id && contract == r.contract, kv_bad_iter,
+                    "Incompatible key-value iterators");
+         if (!r.kv_key) {
+            if (!kv_key)
+               return 0;
+            else
+               return -1;
+         }
+         if (!kv_key)
+            return 1;
+         return compare_blob(*kv_key, *r.kv_key);
+      }
+
+      int32_t kv_it_key_compare(const char* key, uint32_t size) override {
+         if (!kv_key)
+            return 1;
+         return compare_blob(*kv_key, std::string_view{ key, size });
+      }
+
+      kv_it_stat kv_it_move_to_end() override {
+         kv_key.reset();
+         return kv_it_stat::iterator_end;
+      }
+
+      kv_it_stat kv_it_next() override {
+         if (kv_key) {
+            auto it = idx.lower_bound(boost::make_tuple(database_id, contract, *kv_key));
+            if (it != idx.end()) {
+               if (it->kv_key == *kv_key)
+                  ++it;
+               return move_to(it);
+            }
+            kv_key.reset();
+            return kv_it_stat::iterator_end;
+         }
+         return move_to(idx.lower_bound(boost::make_tuple(database_id, contract, prefix)));
+      }
+
+      kv_it_stat kv_it_prev() override {
+         std::decay_t<decltype(idx.end())> it;
+         if (kv_key)
+            it = idx.lower_bound(boost::make_tuple(database_id, contract, *kv_key));
+         else if (next_prefix)
+            it = idx.lower_bound(boost::make_tuple(database_id, contract, *next_prefix));
```

Comment:
This isn't quite right.  Consider:
```
keys = { "\x00\xFF\xFF", "\x01", "\x01\x00" }
prefix = "\x00\xFF\xFF";
next_prefix = "\x01\x00\x00"
```

Then ``lower_bound="\x01\x00"``, decrementing it gives ``"\x01"``, which is outside the prefix, so ``move_to`` returns ``iterator_end``.

Reply:
`lower_bound=end`

Reply:
Oops. Even with that it's not right.

Reply:
Switched `next_prefix` generation.